### PR TITLE
FOGL-9174 Enhancements

### PR DESCRIPTION
**Pull request by ashish-jabble**

### CODE REVIEW

**Review by MarkRiddoch — DISMISSED** [review]

File: python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py [hunk 1]
```diff
@@ -282,19 +338,71 @@ def start(self):
...
-            'timestamp': datetime.fromtimestamp(metric.timestamp, tz=timezone.utc
-                                                ).strftime('%Y-%m-%d %H:%M:%S.%s'),
-            'readings': {metric.name: value}
+            'asset': asset,
+            'timestamp': ts,
+            'readings': readings
         }
         async_ingest.ingest_callback(c_callback, c_ingest_ref, data)
+
+    def validate_topic(self) -> bool:
```
Comment (praveen-garg) [inline python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py:366]: Cannot a user subscribe `spBv1.0/FactoryX/#` ?

File: python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py [hunk 2]
```diff
@@ -43,54 +43,99 @@
...
+        'displayName': 'Topic',
+        'mandatory': 'true',
+        'group': 'Topic'
+    }
 }
 
 c_callback = None
 c_ingest_ref = None
 loop = None
+namespace = "spBv1.0"
```
Comment (praveen-garg) [inline python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py:138]: better to make it `__NAMESPACE = "spBv1.0"` or "NAMESPACE = spBv1.0"

File: python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py [hunk 3]
```diff
@@ -282,19 +338,71 @@ def start(self):
...
+    def validate_topic(self) -> bool:
+        # Split the topic by '/'
+        components = self.topic.split('/')
+
+        # Rule 1: Must start with "spBv1.0"
+        if components[0] != namespace:
+            return False
+
+        # Rule 2: Topic must have 3 or 4 components
+        if len(components) < 4 or len(components) > 5:
```
Comment (praveen-garg) [inline python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py:375]: swap the order.

File: python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py [hunk 4]
```diff
@@ -282,19 +338,71 @@ def start(self):
...
+
+    def validate_topic(self) -> bool:
+        # Split the topic by '/'
+        components = self.topic.split('/')
+
+        # Rule 1: Must start with "spBv1.0"
+        if components[0] != namespace:
+            return False
+
+        # Rule 2: Topic must have 3 or 4 components
```
Comment (praveen-garg) [inline python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py:374]: Topic must have at-least 4  or maximum 5 components, DEVICE_ID is optional

File: python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py [hunk 5]
```diff
@@ -282,19 +338,71 @@ def start(self):
...
+        if len(components) < 4 or len(components) > 5:
+            return False
+
+        # Rule 3: No empty strings allowed in components
+        for component in components[1:]:  # Skip the "spBv1.0"
+            if not component:
```
Comment (praveen-garg) [inline python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py:380]: spaces are allowed, but can a component be just all spaces?

File: python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py [hunk 6]
```diff
@@ -208,27 +253,31 @@ class MqttSubscriberClient(object):
...
         self.mqtt_client = mqtt.Client()
         self.broker_host = config['url']['value']
         self.broker_port = int(config['port']['value'])
         self.username = config['user']['value']
         self.password = config['password']['value']
-        self.topic = config['topic']['value']
         self.asset_name = config['assetName']['value']
+        self.asset_naming = config['assetNaming']['value']
+        self.topic = config['topic']['value']
+        self.topic_fragments = config['topicFragments']['value']
```
Comment (praveen-garg) [inline python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py:267]: May be convert to lower() internally, i.e. Group_id is also acceptable.

File: python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py (hunk 1 above)
Comment (ashish-jabble) [reply python/fledge/plugins/south/mqtt_sparkplug/mqtt_sparkplug.py:371]: Added comments for the validation of a topic; as of now no validation

**Review by MarkRiddoch — APPROVED** [review]

**Pull request merged**